### PR TITLE
Fixed 'Bug 60082 - "Go to Declaration" doesn't work'

### PR DESCRIPTION
**Pull request by mkrueger**
IdString generation was very basic so it failed in almost any advanced case. Unfortunately cecil doesn't have an id string generator implemented so a custom one was needed.

### CODE REVIEW

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/UnresolvedIdStringProvider.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+					for (int i = start; i < parameters.Count; i++){
+						if (i > start)
+							sb.Append (',');
+						AppendTypeName (sb, parameters[i].Type, false, GetTypeParameterCount(entity));
+					}
+					sb.Append (')');
+				}
+			}
+			if (member.SymbolKind == SymbolKind.Operator && (member.Name == "op_Implicit" || member.Name == "op_Explicit")) {
+				sb.Append ('~');
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/UnresolvedIdStringProvider.cs]: Shouldn't we actually be appending `implicit` or `explicit` here?

Comment (mkrueger) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/UnresolvedIdStringProvider.cs]: why ? member name is already appended at line 67.
According to c# language spec. 4.3.1 for implicit & explict operators:

Conversion operators have a trailing “~” followed by the return type.

namespace Acme {
 	class Widget: IProcess 	{ 		
public static explicit operator int(Widget x) {...} 		
public static implicit operator long(Widget x) {...} 	
}
 }

"M:Acme.Widget.op_Explicit(Acme.Widget)~System.Int32" "M:Acme.Widget.op_Implicit(Acme.Widget)~System.Int64"

At that point the string is at:
"M:Acme.Widget.op_Explicit(Acme.Widget)'

this code only appends the ~ suffix with return type.

Y I looked at the spec implementing that - id strings are very complex and I wanted a working version. Even if that code will be thrown away in the future. We should replace the decompiler backend.  I need to ask the ILSpy guys what's up with their rewrite…

Comment (mkrueger) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/UnresolvedIdStringProvider.cs]: ... but even if we replace it we probably need our own id string mapping :/. So having one working implementation isn't the worst thing that can happen.

Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/UnresolvedIdStringProvider.cs]: :+1: sounds good then

**Pull request merged**